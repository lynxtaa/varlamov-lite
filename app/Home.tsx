'use client'

import { useInfiniteQuery, useQueryClient } from '@tanstack/react-query'
import { useEffect, useRef } from 'react'

import ArticleListItem from '../components/ArticleListItem'
import Page from '../components/layouts/Page'
import { useIsOnScreen } from '../hooks/useIsOnScreen'
import { Article } from '../lib/varlamovClient'

async function fetchArticles(lastArticleId: number | undefined) {
	const qs =
		lastArticleId !== undefined
			? new URLSearchParams({ lastArticle: String(lastArticleId) })
			: null
	const response = await fetch(`/api/articles${qs ? `?${qs.toString()}` : ''}`)
	if (!response.ok) {
		throw new Error(`Error requesting ${response.url}: ${response.status}`)
	}
	const articles = (await response.json()) as Article[]
	return articles
}

const staleTime = 30 * 60 * 1000
const cacheTime = 30 * 60 * 1000

export default function Home({ initialData }: { initialData: Article[] }) {
	const queryClient = useQueryClient()

	const { data, error, fetchNextPage, hasNextPage, isFetchingNextPage } =
		useInfiniteQuery<Article[], Error>(
			['articles'],
			({ pageParam }: { pageParam?: number }) =>
				queryClient.fetchQuery(
					['next-articles', pageParam],
					() => fetchArticles(pageParam),
					{ staleTime, cacheTime },
				),
			{
				getNextPageParam: lastPage => lastPage[lastPage.length - 1]?.id,
				staleTime,
				cacheTime,
				initialData: {
					pages: [initialData],
					pageParams: [undefined],
				},
			},
		)

	const buttonRef = useRef<HTMLButtonElement>(null)

	const isButtonVisible = useIsOnScreen(buttonRef.current)

	const articles = data?.pages.flat()

	const lastArticleId = articles?.[articles.length - 1]?.id

	useEffect(() => {
		if (
			lastArticleId !== undefined &&
			!isFetchingNextPage &&
			isButtonVisible &&
			hasNextPage === true
		) {
			void queryClient.prefetchQuery(
				['next-articles', lastArticleId],
				() => fetchArticles(lastArticleId),
				{ staleTime, cacheTime },
			)
		}
	}, [hasNextPage, isButtonVisible, isFetchingNextPage, lastArticleId, queryClient])

	return (
		<Page className="max-w-xl" isHome>
			{articles ? (
				articles.length > 0 ? (
					articles.map(article => <ArticleListItem key={article.id} {...article} />)
				) : (
					<header className="mb-8">
						<h2 className="mb-1 text-xl">Ничего не найдено</h2>
					</header>
				)
			) : error ? (
				<span>{error.message}</span>
			) : null}
			{hasNextPage === true && (
				<button
					className="text-sm py-2 px-3 border rounded border-gray-700 border-solid mb-3 hover:no-underline"
					onClick={() => {
						void fetchNextPage()
					}}
					disabled={isFetchingNextPage}
					ref={buttonRef}
				>
					Показать больше {isFetchingNextPage && '...'}
				</button>
			)}
		</Page>
	)
}
