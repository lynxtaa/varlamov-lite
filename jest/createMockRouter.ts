import { useRouter } from 'next/navigation'
import { vi } from 'vitest'

type AppRouterInstance = ReturnType<typeof useRouter>

export function createMockRouter(
	options?: Partial<AppRouterInstance>,
): AppRouterInstance {
	return {
		push: vi.fn(),
		replace: vi.fn(),
		refresh: vi.fn(),
		prefetch: vi.fn(),
		back: vi.fn(),
		forward: vi.fn(),
		...options,
	}
}
