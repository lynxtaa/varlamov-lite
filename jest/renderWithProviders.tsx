import { render as _render, type RenderOptions } from '@testing-library/react'
import React from 'react'

export const renderWithProviders = (el: React.ReactElement, options?: RenderOptions) =>
	_render(el, { ...options })
