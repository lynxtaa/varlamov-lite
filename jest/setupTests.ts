// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import '@testing-library/jest-dom/extend-expect'
import { clear } from 'jest-date-mock'

import { server } from './server'

jest.mock('next/navigation', () => ({
	useRouter: jest.fn(),
}))

beforeAll(() => {
	server.listen({
		onUnhandledRequest: 'warn',
	})
})

afterEach(() => {
	if (typeof sessionStorage !== 'undefined') {
		sessionStorage.clear()
	}
	if (typeof localStorage !== 'undefined') {
		localStorage.clear()
	}
	clear()

	server.resetHandlers()
})

afterAll(() => {
	server.close()
})

global.IntersectionObserver = class IntersectionObserver {
	observe = jest.fn()
	root = null
	rootMargin = ''
	disconnect = jest.fn()
	thresholds = []
	takeRecords = jest.fn()
	unobserve = jest.fn()
}
