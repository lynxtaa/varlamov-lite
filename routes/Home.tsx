import { useInfiniteQuery } from 'react-query'

import ArticleListItem from '../components/ArticleListItem'
import Page from '../components/layouts/Page'
import { Article } from '../lib/varlamovClient'

async function fetchArticles({ pageParam }: { pageParam?: number }) {
	const qs = pageParam ? new URLSearchParams({ lastArticle: String(pageParam) }) : null
	const response = await fetch(`/api/articles${qs ? `?${qs}` : ''}`)
	if (!response.ok) {
		throw new Error(`Error requesting ${response.url}: ${response.status}`)
	}
	const articles = (await response.json()) as Article[]
	return articles
}

export default function Home() {
	const { data, error, fetchNextPage, hasNextPage, isFetchingNextPage } =
		useInfiniteQuery<Article[], Error>('articles', fetchArticles, {
			getNextPageParam: lastPage => lastPage[lastPage.length - 1]?.id,
			staleTime: 30 * 60 * 1000,
			cacheTime: 30 * 60 * 1000,
		})

	const articles = data && data.pages.flat()

	return (
		<Page className="max-w-xl">
			{articles ? (
				articles.length > 0 ? (
					articles.map(article => <ArticleListItem key={article.id} {...article} />)
				) : (
					<header className="mb-8">
						<h2 className="mb-1 text-xl">Ничего не найдено</h2>
					</header>
				)
			) : error ? (
				<span>{error.message}</span>
			) : null}
			{hasNextPage && (
				<button
					className="text-sm py-2 px-3 border rounded border-gray-700 border-solid mb-3 hover:no-underline"
					onClick={() => fetchNextPage()}
					disabled={isFetchingNextPage}
				>
					Показать больше {isFetchingNextPage && '...'}
				</button>
			)}
		</Page>
	)
}
