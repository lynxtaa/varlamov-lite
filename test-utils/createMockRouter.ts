import { NextRouter } from 'next/router'

export function createMockRouter(options?: Partial<NextRouter>): NextRouter {
	return {
		basePath: '/',
		pathname: '/',
		route: '/',
		query: {},
		asPath: '/',
		push: vi.fn(() => Promise.resolve(true)),
		replace: vi.fn(() => Promise.resolve(true)),
		reload: vi.fn(),
		prefetch: vi.fn(() => Promise.resolve()),
		back: vi.fn(),
		forward: vi.fn(),
		beforePopState: vi.fn(),
		isFallback: false,
		isLocaleDomain: true,
		events: {
			on: vi.fn(),
			off: vi.fn(),
			emit: vi.fn(),
		},
		isReady: true,
		isPreview: false,
		...options,
	}
}
