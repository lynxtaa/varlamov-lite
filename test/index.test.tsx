import { render } from '@testing-library/react'
import { subDays, subHours, subMinutes } from 'date-fns'
import { advanceTo } from 'jest-date-mock'
import { NextRouter, useRouter } from 'next/router'

import { createMockRouter } from '../jest/createMockRouter'
import Home from '../pages/index'

const useRouterMock = useRouter as jest.Mock<NextRouter>

beforeEach(() => {
	useRouterMock.mockReturnValue(createMockRouter())
})

it('shows blog posts', () => {
	const now = new Date('2020-03-20')
	advanceTo(now)

	const container = render(
		<Home
			initialData={[
				{
					id: 1,
					uri: '1.html',
					title: 'Первая новость',
					created_at: subMinutes(now, 2).toISOString(),
				},
				{
					id: 2,
					uri: '2.html',
					title: 'Вторая новость',
					created_at: subHours(now, 2).toISOString(),
				},
				{
					id: 3,
					uri: '3.html',
					title: 'Третья новость',
					created_at: subDays(now, 2).toISOString(),
				},
			]}
		/>,
	)

	expect(container).toMatchSnapshot()
})
