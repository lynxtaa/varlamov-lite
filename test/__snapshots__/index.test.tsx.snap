// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shows blog posts 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="my-0 mx-auto p-4 max-w-xl pb-0"
      >
        <header
          class="flex items-center mt-4 mb-8"
        >
          <h1
            class="text-2xl mr-3 font-bold"
          >
            Блог Ильи Варламова
          </h1>
          <button
            class="ml-auto"
            title="Переключить тему"
            type="button"
          >
            <div
              class="w-7 h-7"
            >
              <svg
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                style="width: 100%; height: 100%;"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M21 12.79A9 9 0 1 1 11.21 3 7 7 0 0 0 21 12.79z"
                />
              </svg>
            </div>
          </button>
        </header>
        <header
          class="mb-8"
        >
          <h2
            class="mb-1 text-xl"
          >
            <a
              class="hover:underline"
              href="/blog/1.html"
            >
              Первая новость
            </a>
          </h2>
          <time
            class="opacity-70"
          >
            сегодня в 2:58
          </time>
        </header>
        <header
          class="mb-8"
        >
          <h2
            class="mb-1 text-xl"
          >
            <a
              class="hover:underline"
              href="/blog/2.html"
            >
              Вторая новость
            </a>
          </h2>
          <time
            class="opacity-70"
          >
            сегодня в 1:00
          </time>
        </header>
        <header
          class="mb-8"
        >
          <h2
            class="mb-1 text-xl"
          >
            <a
              class="hover:underline"
              href="/blog/3.html"
            >
              Третья новость
            </a>
          </h2>
          <time
            class="opacity-70"
          >
            18 марта
          </time>
        </header>
        <div
          class="flex items-center sticky bottom-0 py-4 space-x-6 justify-end bg-gray-50 dark:bg-gray-900"
        >
          <a
            class="hover:underline"
            href="/page/2"
            title="Предыдущая страница"
          >
            <div
              class="w-7 h-7"
            >
              <svg
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                style="width: 100%; height: 100%;"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <line
                  x1="19"
                  x2="5"
                  y1="12"
                  y2="12"
                />
                <polyline
                  points="12 19 5 12 12 5"
                />
              </svg>
            </div>
          </a>
          <span>
            1
          </span>
          <span
            class="opacity-50"
          >
            <div
              class="w-7 h-7"
            >
              <svg
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                style="width: 100%; height: 100%;"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <line
                  x1="5"
                  x2="19"
                  y1="12"
                  y2="12"
                />
                <polyline
                  points="12 5 19 12 12 19"
                />
              </svg>
            </div>
          </span>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="my-0 mx-auto p-4 max-w-xl pb-0"
    >
      <header
        class="flex items-center mt-4 mb-8"
      >
        <h1
          class="text-2xl mr-3 font-bold"
        >
          Блог Ильи Варламова
        </h1>
        <button
          class="ml-auto"
          title="Переключить тему"
          type="button"
        >
          <div
            class="w-7 h-7"
          >
            <svg
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              style="width: 100%; height: 100%;"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M21 12.79A9 9 0 1 1 11.21 3 7 7 0 0 0 21 12.79z"
              />
            </svg>
          </div>
        </button>
      </header>
      <header
        class="mb-8"
      >
        <h2
          class="mb-1 text-xl"
        >
          <a
            class="hover:underline"
            href="/blog/1.html"
          >
            Первая новость
          </a>
        </h2>
        <time
          class="opacity-70"
        >
          сегодня в 2:58
        </time>
      </header>
      <header
        class="mb-8"
      >
        <h2
          class="mb-1 text-xl"
        >
          <a
            class="hover:underline"
            href="/blog/2.html"
          >
            Вторая новость
          </a>
        </h2>
        <time
          class="opacity-70"
        >
          сегодня в 1:00
        </time>
      </header>
      <header
        class="mb-8"
      >
        <h2
          class="mb-1 text-xl"
        >
          <a
            class="hover:underline"
            href="/blog/3.html"
          >
            Третья новость
          </a>
        </h2>
        <time
          class="opacity-70"
        >
          18 марта
        </time>
      </header>
      <div
        class="flex items-center sticky bottom-0 py-4 space-x-6 justify-end bg-gray-50 dark:bg-gray-900"
      >
        <a
          class="hover:underline"
          href="/page/2"
          title="Предыдущая страница"
        >
          <div
            class="w-7 h-7"
          >
            <svg
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              style="width: 100%; height: 100%;"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <line
                x1="19"
                x2="5"
                y1="12"
                y2="12"
              />
              <polyline
                points="12 19 5 12 12 5"
              />
            </svg>
          </div>
        </a>
        <span>
          1
        </span>
        <span
          class="opacity-50"
        >
          <div
            class="w-7 h-7"
          >
            <svg
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              style="width: 100%; height: 100%;"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <line
                x1="5"
                x2="19"
                y1="12"
                y2="12"
              />
              <polyline
                points="12 5 19 12 12 19"
              />
            </svg>
          </div>
        </span>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
